File: packages/sdk-ts/src/wizard/detect.ts
/**
 * Framework / package-manager / DB / auth detection used by the wizard.
 *
 * Spec: gravel-cloud/docs/spec/wizard.md §2 step 1.
 *
 * All detection is best-effort and reversible — we never write anything
 * during detection; we just look.
 */
import { promises as fs } from 'node:fs'
import { join } from 'node:path'

export type PackageManager = 'pnpm' | 'yarn' | 'bun' | 'npm' | 'uv' | 'poetry' | 'pip' | 'pipenv'

export type Framework =
  | 'next-app-router'
  | 'next-pages-router'
  | 'express'
  | 'fastify'
  | 'hono'
  | 'fastapi'
  | 'django'
  | 'flask'
  | 'generic-node'
  | 'generic-asgi'
  | 'generic-wsgi'

export type DbDriver = 'postgres' | 'sqlite' | 'mysql' | 'unknown'

export type AuthProvider = 'clerk' | 'next-auth' | 'better-auth' | 'lucia' | 'auth0' | 'fastapi-users' | 'django-auth' | 'unknown'

export interface DetectionResult {
  cwd: string
  language: 'ts' | 'python'
  packageManager: PackageManager
  framework: Framework
  /**
   * For Next.js projects, where `app/` lives relative to `cwd`. Empty
   * string for the root convention (`./app/`); `'src'` for the src/
   * convention (`./src/app/`); `null` if neither (e.g. pages-router
   * project, or non-Next stack).
   */
  nextAppDir: 'app' | 'src/app' | null
  /**
   * For Next.js projects: true if BOTH `app/` (or `src/app/`) AND
   * `pages/` (or `src/pages/`) exist — incremental migration scenario.
   * The wizard prefers app router but surfaces this so the caller can
   * warn the user.
   */
  nextHasBothRouters: boolean
  database: { driver: DbDriver; envVar: string | null }
  auth: AuthProvider
  existingTracers: string[]
  hasGit: boolean
}

export async function detect(cwd: string = process.cwd()): Promise<DetectionResult> {
  const tsResult = await detectTs(cwd)
  if (tsResult) return tsResult
  const pyResult = await detectPython(cwd)
  if (pyResult) return pyResult
  // No clear language detected — assume TS-generic-node as the safe fallback.
  return {
    cwd,
    language: 'ts',
    packageManager: 'npm',
    framework: 'generic-node',
    nextAppDir: null,
    nextHasBothRouters: false,
    database: { driver: 'unknown', envVar: null },
    auth: 'unknown',
    existingTracers: [],
    hasGit: await pathExists(join(cwd, '.git')),
  }
}

async function detectTs(cwd: string): Promise<DetectionResult | null> {
  const pkgPath = join(cwd, 'package.json')
  if (!(await pathExists(pkgPath))) return null
  const pkg = JSON.parse(await fs.readFile(pkgPath, 'utf8')) as Record<string, any>
  const allDeps = { ...pkg.dependencies, ...pkg.devDependencies }

  const packageManager: PackageManager = (await pathExists(join(cwd, 'pnpm-lock.yaml')))
    ? 'pnpm'
    : (await pathExists(join(cwd, 'yarn.lock')))
      ? 'yarn'
      : // Bun ≥1.2 default is text-format `bun.lock`; older versions used
        // the binary `bun.lockb`. Check both so the wizard catches projects
        // that opted into the new format.
        (await pathExists(join(cwd, 'bun.lock'))) || (await pathExists(join(cwd, 'bun.lockb')))
        ? 'bun'
        : 'npm'

  let framework: Framework = 'generic-node'
  let nextAppDir: 'app' | 'src/app' | null = null
  let nextHasBothRouters = false
  if (allDeps['next']) {
    // Detect the App Router by looking in both conventional locations:
    // `app/` at the repo root, or `src/app/`. Same for `pages/`. A project
    // with both directories (incremental migration) is flagged via
    // `nextHasBothRouters` so the wizard surfaces a warning rather than
    // silently picking app-router.
    const hasAppRoot = await pathExists(join(cwd, 'app'))
    const hasAppSrc = await pathExists(join(cwd, 'src', 'app'))
    const hasPagesRoot = await pathExists(join(cwd, 'pages'))
    const hasPagesSrc = await pathExists(join(cwd, 'src', 'pages'))
    if (hasAppRoot) nextAppDir = 'app'
    else if (hasAppSrc) nextAppDir = 'src/app'
    framework = nextAppDir ? 'next-app-router' : 'next-pages-router'
    nextHasBothRouters = !!nextAppDir && (hasPagesRoot || hasPagesSrc)
  } else if (allDeps['express']) {
    framework = 'express'
  } else if (allDeps['fastify']) {
    framework = 'fastify'
  } else if (allDeps['hono']) {
    framework = 'hono'
  }

  const auth: AuthProvider = allDeps['@clerk/nextjs']
    ? 'clerk'
    : allDeps['@clerk/clerk-js']
      ? 'clerk'
      : allDeps['next-auth']
        ? 'next-auth'
        : allDeps['better-auth']
          ? 'better-auth'
          : allDeps['lucia']
            ? 'lucia'
            : allDeps['@auth0/auth0-react'] || allDeps['@auth0/nextjs-auth0']
              ? 'auth0'
              : 'unknown'

  const dbEnv = await readEnvVar(cwd, ['DATABASE_URL', 'POSTGRES_URL', 'NEON_DATABASE_URL'])
  const database = inferDb(dbEnv)

  const existingTracers: string[] = []
  if (allDeps['@sentry/node'] || allDeps['@sentry/nextjs']) existingTracers.push('Sentry')
  if (allDeps['langsmith']) existingTracers.push('LangSmith')
  if (allDeps['langfuse']) existingTracers.push('Langfuse')

  return {
    cwd,
    language: 'ts',
    packageManager,
    framework,
    nextAppDir,
    nextHasBothRouters,
    database,
    auth,
    existingTracers,
    hasGit: await pathExists(join(cwd, '.git')),
  }
}

async function detectPython(cwd: string): Promise<DetectionResult | null> {
  const hasPyproject = await pathExists(join(cwd, 'pyproject.toml'))
  const hasManagePy = await pathExists(join(cwd, 'manage.py'))
  const hasReqs = await pathExists(join(cwd, 'requirements.txt'))
  if (!hasPyproject && !hasManagePy && !hasReqs) return null

  const pmText = await readManyOptional(cwd, ['pyproject.toml', 'requirements.txt', 'Pipfile'])
  const packageManager: PackageManager = (await pathExists(join(cwd, 'uv.lock')))
    ? 'uv'
    : (await pathExists(join(cwd, 'poetry.lock')))
      ? 'poetry'
      : (await pathExists(join(cwd, 'Pipfile.lock')))
        ? 'pipenv'
        : 'pip'

  const allText = pmText.join('\n').toLowerCase()
  let framework: Framework = 'generic-asgi'
  if (hasManagePy || allText.includes('django')) framework = 'django'
  else if (allText.includes('fastapi')) framework = 'fastapi'
  else if (allText.includes('flask')) framework = 'flask'

  const auth: AuthProvider = allText.includes('django.contrib.auth') || hasManagePy
    ? 'django-auth'
    : allText.includes('fastapi-users')
      ? 'fastapi-users'
      : 'unknown'

  const dbEnv = await readEnvVar(cwd, ['DATABASE_URL', 'POSTGRES_URL'])
  const database = inferDb(dbEnv)

  const existingTracers: string[] = []
  if (allText.includes('sentry-sdk')) existingTracers.push('Sentry')
  if (allText.includes('langsmith')) existingTracers.push('LangSmith')
  if (allText.includes('langfuse')) existingTracers.push('Langfuse')

  return {
    cwd,
    language: 'python',
    packageManager,
    framework,
    nextAppDir: null,
    nextHasBothRouters: false,
    database,
    auth,
    existingTracers,
    hasGit: await pathExists(join(cwd, '.git')),
  }
}

function inferDb(envValue: { name: string; value: string } | null): DetectionResult['database'] {
  if (!envValue) return { driver: 'unknown', envVar: null }
  if (envValue.value.startsWith('postgres')) return { driver: 'postgres', envVar: envValue.name }
  if (envValue.value.startsWith('mysql')) return { driver: 'mysql', envVar: envValue.name }
  if (envValue.value.startsWith('file:') || envValue.value.endsWith('.db')) {
    return { driver: 'sqlite', envVar: envValue.name }
  }
  return { driver: 'unknown', envVar: envValue.name }
}

async function readEnvVar(cwd: string, candidates: string[]): Promise<{ name: string; value: string } | null> {
  for (const file of ['.env.local', '.env']) {
    const path = join(cwd, file)
    if (!(await pathExists(path))) continue
    const text = await fs.readFile(path, 'utf8')
    for (const line of text.split('\n')) {
      const m = /^\s*(\w+)\s*=\s*(.+?)\s*$/.exec(line)
      if (!m) continue
      const [, name, raw] = m as unknown as [string, string, string]
      if (!candidates.includes(name)) continue
      const value = raw.replace(/^['"]|['"]$/g, '')
      return { name, value }
    }
  }
  return null
}

async function readManyOptional(cwd: string, files: string[]): Promise<string[]> {
  const results: string[] = []
  for (const f of files) {
    try {
      results.push(await fs.readFile(join(cwd, f), 'utf8'))
    } catch {
      /* ignore */
    }
  }
  return results
}

async function pathExists(path: string): Promise<boolean> {
  try {
    await fs.access(path)
    return true
  } catch {
    return false
  }
}

File: python/gravel/src/artanis_gravel/wizard/detect.py
"""Framework / package-manager / DB / auth detection. Parity with src/wizard/detect.ts."""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

PackageManager = Literal["pnpm", "yarn", "bun", "npm", "uv", "poetry", "pip", "pipenv"]
Framework = Literal[
    "next-app-router", "next-pages-router", "express", "fastify", "hono",
    "fastapi", "django", "flask", "generic-node", "generic-asgi", "generic-wsgi",
]
DbDriver = Literal["postgres", "sqlite", "mysql", "unknown"]
AuthProvider = Literal[
    "clerk", "next-auth", "better-auth", "lucia", "auth0",
    "fastapi-users", "django-auth", "unknown",
]


@dataclass
class DetectionResult:
    cwd: str
    language: Literal["ts", "python"]
    package_manager: PackageManager
    framework: Framework
    database: dict  # {"driver": DbDriver, "envVar": str | None}
    auth: AuthProvider
    existing_tracers: list[str] = field(default_factory=list)
    has_git: bool = False
    # Next.js: where `app/` lives. "app" for root, "src/app" for src/, None
    # if neither (pages-only, or non-Next).
    next_app_dir: Literal["app", "src/app", None] = None
    # Next.js: True if both app/ (or src/app/) AND pages/ (or src/pages/)
    # exist — incremental migration scenario.
    next_has_both_routers: bool = False


def detect(cwd: str | Path | None = None) -> DetectionResult:
    cwd = Path(cwd) if cwd else Path.cwd()

    py_result = _detect_python(cwd)
    if py_result is not None:
        return py_result

    ts_result = _detect_ts(cwd)
    if ts_result is not None:
        return ts_result

    return DetectionResult(
        cwd=str(cwd),
        language="ts",
        package_manager="npm",
        framework="generic-node",
        database={"driver": "unknown", "envVar": None},
        auth="unknown",
        existing_tracers=[],
        has_git=(cwd / ".git").exists(),
    )


def _detect_python(cwd: Path) -> DetectionResult | None:
    has_pyproject = (cwd / "pyproject.toml").exists()
    has_manage = (cwd / "manage.py").exists()
    has_reqs = (cwd / "requirements.txt").exists()
    if not (has_pyproject or has_manage or has_reqs):
        return None

    pm: PackageManager
    if (cwd / "uv.lock").exists():
        pm = "uv"
    elif (cwd / "poetry.lock").exists():
        pm = "poetry"
    elif (cwd / "Pipfile.lock").exists():
        pm = "pipenv"
    else:
        pm = "pip"

    text = ""
    for f in ("pyproject.toml", "requirements.txt", "Pipfile"):
        p = cwd / f
        if p.exists():
            text += p.read_text(encoding="utf-8", errors="ignore") + "\n"
    text_lower = text.lower()

    framework: Framework = "generic-asgi"
    if has_manage or "django" in text_lower:
        framework = "django"
    elif "fastapi" in text_lower:
        framework = "fastapi"
    elif "flask" in text_lower:
        framework = "flask"

    auth: AuthProvider = (
        "django-auth" if (has_manage or "django.contrib.auth" in text_lower)
        else "fastapi-users" if "fastapi-users" in text_lower
        else "unknown"
    )

    db_env = _read_env_var(cwd, ["DATABASE_URL", "POSTGRES_URL"])
    database = _infer_db(db_env)

    existing: list[str] = []
    if "sentry-sdk" in text_lower:
        existing.append("Sentry")
    if "langsmith" in text_lower:
        existing.append("LangSmith")
    if "langfuse" in text_lower:
        existing.append("Langfuse")

    return DetectionResult(
        cwd=str(cwd),
        language="python",
        package_manager=pm,
        framework=framework,
        database=database,
        auth=auth,
        existing_tracers=existing,
        has_git=(cwd / ".git").exists(),
    )


def _detect_ts(cwd: Path) -> DetectionResult | None:
    pkg = cwd / "package.json"
    if not pkg.exists():
        return None
    import json
    p = json.loads(pkg.read_text(encoding="utf-8"))
    deps = {**p.get("dependencies", {}), **p.get("devDependencies", {})}

    pm: PackageManager
    if (cwd / "pnpm-lock.yaml").exists():
        pm = "pnpm"
    elif (cwd / "yarn.lock").exists():
        pm = "yarn"
    elif (cwd / "bun.lock").exists() or (cwd / "bun.lockb").exists():
        # Bun ≥1.2 default is text-format `bun.lock`; older was binary
        # `bun.lockb`. Accept either.
        pm = "bun"
    else:
        pm = "npm"

    framework: Framework = "generic-node"
    next_app_dir: Literal["app", "src/app", None] = None
    next_has_both_routers = False
    if "next" in deps:
        has_app_root = (cwd / "app").exists()
        has_app_src = (cwd / "src" / "app").exists()
        has_pages_root = (cwd / "pages").exists()
        has_pages_src = (cwd / "src" / "pages").exists()
        if has_app_root:
            next_app_dir = "app"
        elif has_app_src:
            next_app_dir = "src/app"
        framework = "next-app-router" if next_app_dir else "next-pages-router"
        next_has_both_routers = bool(next_app_dir) and (has_pages_root or has_pages_src)
    elif "express" in deps:
        framework = "express"
    elif "fastify" in deps:
        framework = "fastify"
    elif "hono" in deps:
        framework = "hono"

    auth: AuthProvider = "unknown"
    if any(k.startswith("@clerk/") for k in deps):
        auth = "clerk"
    elif "next-auth" in deps:
        auth = "next-auth"
    elif "better-auth" in deps:
        auth = "better-auth"
    elif "lucia" in deps:
        auth = "lucia"
    elif any(k.startswith("@auth0/") for k in deps):
        auth = "auth0"

    db_env = _read_env_var(cwd, ["DATABASE_URL", "POSTGRES_URL", "NEON_DATABASE_URL"])
    database = _infer_db(db_env)

    existing: list[str] = []
    if any(k.startswith("@sentry/") for k in deps):
        existing.append("Sentry")
    if "langsmith" in deps:
        existing.append("LangSmith")
    if "langfuse" in deps:
        existing.append("Langfuse")

    return DetectionResult(
        cwd=str(cwd),
        language="ts",
        package_manager=pm,
        framework=framework,
        database=database,
        auth=auth,
        existing_tracers=existing,
        has_git=(cwd / ".git").exists(),
        next_app_dir=next_app_dir,
        next_has_both_routers=next_has_both_routers,
    )


def _read_env_var(cwd: Path, candidates: list[str]) -> tuple[str, str] | None:
    for f in (".env.local", ".env"):
        p = cwd / f
        if not p.exists():
            continue
        for line in p.read_text(encoding="utf-8").splitlines():
            m = re.match(r"\s*(\w+)\s*=\s*(.+?)\s*$", line)
            if not m:
                continue
            name, raw = m.group(1), m.group(2)
            if name not in candidates:
                continue
            value = raw.strip("'\"")
            return name, value
    return None


def _infer_db(env: tuple[str, str] | None) -> dict:
    if env is None:
        return {"driver": "unknown", "envVar": None}
    name, value = env
    if value.startswith("postgres"):
        return {"driver": "postgres", "envVar": name}
    if value.startswith("mysql"):
        return {"driver": "mysql", "envVar": name}
    if value.startswith("file:") or value.endswith(".db"):
        return {"driver": "sqlite", "envVar": name}
    return {"driver": "unknown", "envVar": name}
